src/main/java/com/yahoo/memory/CompareAndCopy.java

```diff
@@ -0,0 +1,134 @@
+        if (thisByte > thatByte) { return  1; }
+      }
+    }
+    return Long.compare(thisLengthBytes, thatLengthBytes);
+  }
+
+  static void copy(final ResourceState srcState, final long srcOffsetBytes,
+      final ResourceState dstState, final long dstOffsetBytes, final long lengthBytes) {
+    srcState.checkValid();
+    checkBounds(srcOffsetBytes, lengthBytes, srcState.getCapacity());
+    dstState.checkValid();
+    checkBounds(dstOffsetBytes, lengthBytes, dstState.getCapacity());
+    final long srcAdd = srcState.getCumBaseOffset() + srcOffsetBytes;
+    final long dstAdd = dstState.getCumBaseOffset() + dstOffsetBytes;
+    copyMemory(srcState.getUnsafeObject(), srcAdd, dstState.getUnsafeObject(), dstAdd,
+        lengthBytes);
+  }
+
+  //Used by all of the get/put array methods in Buffer and Memory classes
+  static final void copyMemoryCheckingDifferentObject(final Object srcUnsafeObj,
+      final long srcAdd, final Object dstUnsafeObj, final long dstAdd, final long lengthBytes) {
+    if (srcUnsafeObj != dstUnsafeObj) {
+      copyNonOverlappingMemoryWithChunking(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd,
+          lengthBytes);
+    } else {
+      throw new IllegalArgumentException("Not expecting to copy to/from array which is the "
+          + "underlying object of the memory at the same time");
+    }
+  }
+
+  //only valid and bounds checks have been performed at this point
+  private static void copyMemory(final Object srcUnsafeObj, final long srcAdd,
+      final Object dstUnsafeObj, final long dstAdd, final long lengthBytes) {
+    if (srcUnsafeObj != dstUnsafeObj) {
+      //either srcArray != dstArray OR one of them is off-heap
+      copyNonOverlappingMemoryWithChunking(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd,
+          lengthBytes);
+    } else { //either srcArray == dstArray OR both src and dst are off-heap
+      copyMemoryOverlapAddressCheck(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd, lengthBytes);
+    }
+  }
+
+  /**
+   * At this point either srcArray == dstArray OR both src and dst are off-heap.
+   * Performs overlapping address check. If addresses do not overlap, proceed to
+   * {@link #copyNonOverlappingMemoryWithChunking(Object, long, Object, long, long)}; otherwise
+   * fall back on <i>Unsafe.copyMemory(...)</i> tolerating potentially long
+   * Time to Safe Point pauses.
+   * If srcAdd == dstAdd an exception will be thrown.
+   * @param srcUnsafeObj The source array object, it may be null.
+   * @param srcAdd The cumulative source offset
+   * @param dstUnsafeObj The destination array object, it may be null
+   * @param dstAdd The cumulative destination offset
+   * @param lengthBytes The length to be copied in bytes
+   */
+  private static void copyMemoryOverlapAddressCheck(final Object srcUnsafeObj, final long srcAdd,
+      final Object dstUnsafeObj, final long dstAdd, final long lengthBytes) {
+    if (((srcAdd + lengthBytes) <= dstAdd) || ((dstAdd + lengthBytes) <= srcAdd)) {
```

Comment:
BTW addresses could be negative (Java's `long` is always signed), so we must use overflow-aware comparisons like `a - b <= 0` instead of `a <= b`.

Reply:
Or maybe not, needs to research

Reply:
Good catch.

Reply:
Sorry, it was dumb. All user-space addresses are positive. I've reverted your commit.

Reply:
Do you have any other concerns about this PR?

Reply:
Need to add equals(...) and fix the java version so that it works for 1.7 and >= 1.8.  If the user needs to add JVM args to allow illegal access, so be it, for now.

Reply:
It could be done in separate PRs.